File: pgx-kb/src/library-service/library-service.js
import {knowledgeSet1, knowledgeSet2, knowledgeSet3} from "./knowledge-sets.js";

async function run(input) {

  try {
    const intermediateResults = await Promise.all(
      knowledgeSet1.map(f => f(input['diplotype']))
    );

    const responseGenes = new Set(intermediateResults.map(obj => Object.keys(obj)[0]));

    // Add missing genes
    for (const gene in input['diplotype']) {
      if (!responseGenes.has(gene)) {
        const value = input['diplotype'][gene];
        if (value) {
          intermediateResults.push({ [gene]: { diplotype: value, phenotype: "" } });
        } else {
          intermediateResults.push({ [gene]: {} });
        }
      }
    }

    const mergedResults = intermediateResults.reduce((acc, obj) => {
      return { ...acc, ...obj };
    }, {});
    const finalResultsKS2 = await Promise.all(
      knowledgeSet2.map(f => f(mergedResults))
    );

    const finalResultsKS3 = await Promise.all(
      knowledgeSet3.map(f => f(mergedResults))
    );
    return {
      intermediate: mergedResults,
      finalKS2: finalResultsKS2,
      finalKS3: finalResultsKS3
    };
  } catch (err) {
    console.error(err);
    return { error: 'Execution failed' };
  }
}


export default { run };

File: pgx-kb/pgx_kb/library_service.py
from .knowledge_sets import knowledgeSet1Payloads, knowledgeSet2Payloads

def run(input_data):
    try:
        # Run all functions in knowledgeSet1Payloads concurrently
        intermediate_results = [f(input_data['diplotype']) for f in knowledgeSet1Payloads]

        merged_results = {}
        for obj in intermediate_results:
            merged_results.update(obj)

        # Run all functions in knowledgeSet2Payloads concurrently
        final_results_ks2 = [f(merged_results) for f in knowledgeSet2Payloads]
        print({
            "intermediate": merged_results,
            "finalKS2": final_results_ks2,
        })
        return {
            "intermediate": merged_results,
            "finalKS2": final_results_ks2,
        }
    except Exception as e:
        print(f"Error: {e}")
        
# run({"diplotype": {"CYP2D6": "*3/*3"}})